File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;

public class Main {
    static int R, C;
    static char[][] board;
    static HashSet<Character> words = new HashSet<>();
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static int answer = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] RC = br.readLine().split(" ");
        R = Integer.parseInt(RC[0]);
        C = Integer.parseInt(RC[1]);

        board = new char[R][C];
        for (int i = 0; i < R; i++) {
            String row = br.readLine();
            for (int j = 0; j < C; j++) {
                board[i][j] = row.charAt(j);
            }
        }

        words.add(board[0][0]);
        dfs(0, 0, 1);
        System.out.println(answer);
    }

    static void dfs(int x, int y, int cnt) {
        answer = Math.max(answer, cnt);

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx >= 0 && nx < R && ny >= 0 && ny < C && !words.contains(board[nx][ny])) {
                words.add(board[nx][ny]);
                dfs(nx, ny, cnt + 1);
                words.remove(board[nx][ny]);
            }
        }
    }
}

File: 백준/Gold/1987. 알파벳/알파벳.py
# 지금까지 지나온 모든 칸에 적혀있는 알파벳과 달라야 한다.
# 말이 최대 몇 칸을 지날 수 있는지?
import sys
R, C = map(int, sys.stdin.readline().split())
board = [list(sys.stdin.readline().rstrip()) for _ in range(R)]
words = set()
dx = [1,-1,0,0]
dy = [0,0,1,-1]

answer = 0
def dfs(x, y,cnt) :
    global answer
    answer = max(answer, cnt)
    for i in range(4) :
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < R and 0 <= ny < C and board[nx][ny] not in words :
                words.add(board[nx][ny])
                dfs(nx,ny,cnt+1)
                words.remove(board[nx][ny])
words.add(board[0][0])
dfs(0,0,1)
print(answer)

